File: Net/SSOExtension/VersionOne.SDK.APIClient.SSOExtension/V1SsoConnector.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Web;

namespace VersionOne.SDK.APIClient.SSOExtension
{
    /// <summary>
    /// IAPIConnector that supports SSO authentication
    /// </summary>
    public class V1SsoConnector : IAPIConnector
    {
        #region Attributes
        private readonly string _v1Url;
        private readonly string _idpUrl;
        private readonly string _username;
        private readonly string _password;
        private readonly string _authenticationUri;
        private readonly bool _integratedauth;
        private readonly ResponseParserFactory _responseParserFactory;
        #endregion

        /// <summary>
        /// Helper method to create a Meta connection (ServerUrl + /meta.v1/)
        /// </summary>
        /// <param name="config">configuration</param>
        /// <returns></returns>
        public static V1SsoConnector CreateMetaConnection(IVersionOneSsoConfiguration config) { return new V1SsoConnector(config, "/meta.v1/"); }

        /// <summary>
        /// Helper method to create a data connection (ServerUrl + /rest-1.v1/)
        /// </summary>
        /// <param name="config">configuration</param>
        /// <param name="authenticate">true if this method should attempt authentication</param>
        /// <returns></returns>
        public static V1SsoConnector CreateDataConnection(IVersionOneSsoConfiguration config, bool authenticate)
        {
            var dataConnector = new V1SsoConnector(config, "/rest-1.v1/");
            if(authenticate)
                dataConnector.Authenticate();
            return dataConnector;
        }

        /// <summary>
        /// Construction
        /// </summary>
        /// <param name="config">configuration object</param>
        /// <param name="pathExtension">path extension (/meta.v1/ or /rest-1.v1/)</param>
        public V1SsoConnector(IVersionOneSsoConfiguration config, string pathExtension)
        {
            _v1Url = config.ServerUrl + pathExtension;
            _idpUrl = config.IdpUrl;
            _username = config.Username;
            _password = config.Password;
            _integratedauth = config.IntegratedAuth;
            _authenticationUri = config.AuthenticationUri;
            _responseParserFactory = new ResponseParserFactory(config);
        }

        /// <summary>
        /// Authenticates this connection.  While not required, it is recommended that you call
        /// this method for the "Services" connection ensure that you can authenticate before 
        /// running your code.  
        /// 
        /// This method is called automatically if the VersionOne server requires authentication.
        /// </summary>
        public void Authenticate()
        {
            var idpResponse = IdentityProviderRequest();
            var spResponse = IdentityProviderAuthentication(idpResponse);
            ServiceProviderRequest(spResponse);
        }

        #region Implementation of IAPIConnector Methods
        public Stream GetData()
        {
            return GetData(string.Empty);
        }

        public Stream GetData(string path)
        {
            HttpWebRequest request = CreateRequest(_v1Url + path);
            WebResponse response = request.GetResponse();
            if (((HttpWebResponse)response).StatusCode == HttpStatusCode.Redirect)
            {
                // ToDo: Need some retry logic here so we don't get caught in a loop
                Authenticate();
                return GetData(path);
            }

            #region Code used when debugging
            //Console.WriteLine("");
            //Console.WriteLine("vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv");
            //Console.WriteLine("Get....");
            //Console.WriteLine("URL: " + url);
            //Console.WriteLine("Response from: " + response.ResponseUri);
            //Console.WriteLine("Status: " + response.StatusCode);
            //Console.WriteLine("---------------------------------------------------------------------------------------------------");
            //Console.WriteLine(response.Headers.ToString());
            //Console.WriteLine("---------------------------------------------------------------------------------------------------");
            //Console.WriteLine(response.ToString());
            //Console.WriteLine("^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^");))
            #endregion
            
            return response.GetResponseStream();
        }

        public Stream SendData(string path, string data)
        {
            HttpWebRequest request = CreateRequest(_v1Url + path);
            request.ServicePoint.Expect100Continue = false;

            request.Method = "POST";
            request.ContentType = "type/xml";

            #region Code used when debugging
            //Console.WriteLine("");
            //Console.WriteLine("vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv");
            //Console.WriteLine("POST....");
            //Console.WriteLine("URL: " + _v1Url + path);
            //Console.WriteLine(request.Headers.ToString());
            //Console.WriteLine(data);
            //Console.WriteLine("^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^");
            //Console.WriteLine("");
            #endregion

            using (var stream = new StreamWriter(request.GetRequestStream()))
                stream.Write(data);

            return request.GetResponse().GetResponseStream();

        }

        private readonly IDictionary<string, HttpWebRequest> _requests = new Dictionary<string, HttpWebRequest>();
        public Stream BeginRequest(string path)
        {
            HttpWebRequest req = CreateRequest(_v1Url + path);
            _requests[path] = req;
            req.Method = "POST";
            return req.GetRequestStream();
        }

        public Stream EndRequest(string path, string contentType)
        {
            HttpWebRequest req = _requests[path];
            _requests.Remove(path);
            req.ContentType = contentType;
            return req.GetResponse().GetResponseStream();
        }

        private readonly IDictionary<string, string> _customHttpHeaders = new Dictionary<string, string>();
        public IDictionary<string, string> CustomHttpHeaders
        {
            get { return _customHttpHeaders; }
        }

        #endregion

        /// <summary>
        /// Create a HTTPWebREquest
        /// </summary>
        /// <param name="path"></param>
        /// <returns></returns>
        private HttpWebRequest CreateRequest(string path)
        {
            var request = (HttpWebRequest)WebRequest.Create(path);

            request.Headers.Add("Accept-Language", System.Globalization.CultureInfo.CurrentCulture.Name);

            foreach (KeyValuePair<string, string> pair in _customHttpHeaders)
            {
                request.Headers.Add(pair.Key, pair.Value);
            }
            request.CookieContainer = CookieJar;
            request.Credentials = Credentials;
            request.UnsafeAuthenticatedConnectionSharing = true;
            request.AllowAutoRedirect = false;
            return request;
        }

        #region CredentialCache
        private CredentialCache _credentials;
        private CredentialCache Credentials
        {
            get
            {
                if (_credentials == null)
                {
                    _credentials = new CredentialCache();
                    var uri = new Uri(_authenticationUri);

                    if (!string.IsNullOrEmpty(_username) || !string.IsNullOrEmpty(_password))
                    {
                        var credential = new NetworkCredential(_username, _password);
                        if (_integratedauth)
                        {
                            _credentials.Add(uri, "Negotiate", credential);
                            _credentials.Add(uri, "NTLM", credential);
                        }
                        else
                            _credentials.Add(uri, "Basic", credential);
                    }
                    else
                    {
                        var credential = CredentialCache.DefaultCredentials as NetworkCredential;
                        if (credential != null)
                        {
                            _credentials.Add(uri, "Negotiate", credential);
                            _credentials.Add(uri, "NTLM", credential);
                        }
                    }
                }
                return _credentials;
            }
        }
        #endregion

        #region CookieContainer
        private CookieContainer _cookieJar;
        private CookieContainer CookieJar
        {
            get { return _cookieJar ?? (_cookieJar = new CookieContainer()); }
        }
        #endregion

        #region Authentication Private Methods
        /// <summary>
        /// Make a request to the IdP and return the response.
        /// 
        /// ** This is here, as opposed to a seperate object because we need to share the Cookie container **
        /// 
        /// </summary>
        /// <returns>response from IDP server</returns>
        private IResponseParser IdentityProviderRequest()
        {
            HttpWebRequest idpRequest = CreateRequest(_idpUrl);
            idpRequest.AllowAutoRedirect = true;

            var responseParser = _responseParserFactory.CreateIdentityProviderResponseParser();

            using (WebResponse idpResponse = idpRequest.GetResponse())
            {
                using (Stream stream = idpResponse.GetResponseStream())
                    responseParser.LoadResponse(stream);

                responseParser.UrlAuthority = idpResponse.ResponseUri.GetLeftPart(UriPartial.Authority);
            }
            return responseParser;
        }

        /// <summary>
        /// Post the response from the Identity Provider to the URL specified for authentication
        /// 
        /// ** This is here, as opposed to a seperate object because we need to share the Cookie container **
        /// 
        /// </summary>
        /// <param name="response">initial response from IDP server (before authentication)</param>
        /// <returns>response from IDP server after authentication</returns>
        private IResponseParser IdentityProviderAuthentication(IResponseParser response)
        {
            var formData = response.PostData;
            formData.SetCredentials(_username, _password);

            var postData = formData.ToString();

            // Prepare web request...
            HttpWebRequest request = CreateRequest(response.PostUrl);
            request.AllowAutoRedirect = true;
            request.ServicePoint.Expect100Continue = false;
            request.Method = "POST";
            request.ContentType = "application/x-www-form-urlencoded";
            request.ContentLength = postData.Length;
            using (var stream = new StreamWriter(request.GetRequestStream()))
            {
                stream.Write(postData);
            }

            var responseParser = _responseParserFactory.CreateServiceProviderResponseParser();
            using (WebResponse idpResponse = request.GetResponse())
            {
                using (Stream responseStream = idpResponse.GetResponseStream())
                    responseParser.LoadResponse(responseStream);
            }
            return responseParser;
        }

        /// <summary>
        /// Post data to the service provider
        /// </summary>
        /// <param name="response">response from IDP after authentication</param>
        private void ServiceProviderRequest(IResponseParser response)
        {
            var formData = response.PostData;
            var postData = formData.ToString();

            // Prepare web request...
            string url = HttpUtility.HtmlDecode(response.PostUrl);
            HttpWebRequest request = CreateRequest(url);
            request.AllowAutoRedirect = true;
            request.ServicePoint.Expect100Continue = false;
            request.Method = "POST";
            request.ContentType = "application/x-www-form-urlencoded";
            request.ContentLength = postData.Length;

            using (var stream = new StreamWriter(request.GetRequestStream()))
            {
                stream.Write(postData);
            }
            request.GetResponse().Close();
        }

        #endregion

    }
}

File: Java/VersionOneSSO/src/com/versionone/sso/V1SsoConnector.java
package com.versionone.sso;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.Reader;
import java.net.Authenticator;
import java.net.MalformedURLException;
import java.net.PasswordAuthentication;
import java.util.HashMap;
import java.util.Map;

import org.apache.commons.lang.StringEscapeUtils;

import sun.net.www.protocol.http.AuthCacheImpl;
import sun.net.www.protocol.http.AuthCacheValue;

import com.versionone.apiclient.ConnectionException;
import com.versionone.apiclient.IAPIConnector;

/**
 * IAPIConnector that supports SSO authenticationZ
 * @author jerry
 *
 */
public class V1SsoConnector implements IAPIConnector {

    private String _v1Url;
    private String _idpUrl;
    private String _username;
    private String _password;
        
	/**
	 * Additional headers for request to the VersionOne server
	 */
	public final Map<String, String> customHttpHeaders = new HashMap<String, String>();

	/**
	 * Helper method to create a Meta connection (ServerUrl + /meta.v1/)
	 * @param config - Configuration 
	 * @return
	 */
	public static V1SsoConnector createMetaConnector(IVersionOneSSOConfiguration config)
	{
		return new V1SsoConnector(config, "/meta.v1/"); 
	}
	
	/**
	 * Helper method to create a data connection (ServerUrl + /rest-1.v1/)
	 * @param config - Configuration 
	 * @param authenticate - true if this method should attempt authentication
	 * @return
	 * @throws ConnectionException
	 */
	public static V1SsoConnector createDataConnector(IVersionOneSSOConfiguration config, boolean authenticate) throws ConnectionException
	{
		V1SsoConnector dataConnector = new V1SsoConnector(config, "/rest-1.v1/");
        if(authenticate)
            dataConnector.authenticate();
        return dataConnector;
	}

	/**
	 * Construction
	 * @param config - configuration
	 * @param path - path extension (/meta.v1/ or /rest-1.v1/)
	 */
	public V1SsoConnector(IVersionOneSSOConfiguration config, String path)
	{
	    _v1Url = config.getServerUrl() + path;
	    _username = config.getUsername();
	    _password = config.getPassword();
	    
	    _idpUrl = config.getIdpUrl();
	    ResponseParserFactory.idpResponseParser = config.getIdpResponseParser();
	    ResponseParserFactory.spResponseParser = config.getSpResponseParser();
	    
		// WORKAROUND: http://bugs.sun.com/bugdatabase/view_bug.do?bug_id=6626700
		if (_username != null) {
			AuthCacheValue.setAuthCache(new AuthCacheImpl());
			Authenticator.setDefault(new Credentials(_username, _password));
		}    
	}

	/**
	 * Authenticate the SSO Connection
	 * @throws ConnectionException
	 */
    public void authenticate() throws ConnectionException  
    {  	
        IResponseParser idpResponse = IdentityProviderRequest();
        IResponseParser spResponse = IdentityProviderAuthentication(idpResponse);
        ServiceProviderRequest(spResponse);       
    }

    /**
     * Make a request to the IdP to get the Response value.
     * 
     * ** This is here, as opposed to a separate object because we need to share the Cookie container **
     * 
     * @return
     * @throws ConnectionException
     */
    private IResponseParser IdentityProviderRequest() throws ConnectionException 
    {
    	IResponseParser idpResponseParser = ResponseParserFactory.CreateIdentityProviderResponseParser(); 	
    	HttpGet request;
		try {
			request = HttpGet.create(StringEscapeUtils.unescapeXml(_idpUrl));		
		} catch (MalformedURLException e) {
			throw new ConnectionException("Invalid IdP URL " + _idpUrl, e);
		}
		
    	HttpResponse response = new HttpClient().execute(request);     	
    	
    	try {
			idpResponseParser.loadResponse(response.getInputStream());
			idpResponseParser.setUrlAuthority(response.getResponseAuthority());
		} catch (IOException e) {
			throw new ConnectionException("Error Reading IPD Response",e);	
		} catch (Exception e) {
			throw new ConnectionException("Error getting response authority", e);
		}
    	
    	return idpResponseParser;    	

    }

	/**
	 * Post the response from the Identity Provider challenge to the proper URL for authentication
	 * 
	 * ** This is here, as opposed to a separate object because we need to share the Cookie container **
	 * 
	 * @param idpResponse - The response parser from the Identity Provider request 
	 * @return the response parser necessary to make the Service Provider request
	 * @throws ConnectionException
	 */
	private IResponseParser IdentityProviderAuthentication(IResponseParser idpResponse) throws ConnectionException
    {
        IResponseParser spResponseParser = ResponseParserFactory.CreateServiceProviderResponseParser();
        
    	PostData formData = idpResponse.getPostData();
    	formData.setCredentials(_username, _password);
    	
        // Prepare web request...
        HttpPost request;
		try {

			request = HttpPost.create(idpResponse.getPostUrl(), "application/x-www-form-urlencoded");
	        request.setData(formData.toString());

		} catch (MalformedURLException e) {
			throw new ConnectionException("Invalid URL from IDP", e);
		}
        
        HttpResponse spResponse = new HttpClient().execute(request);
        
        try {
			spResponseParser.loadResponse(spResponse.getInputStream());
		} catch (Exception e) {
			throw new ConnectionException("Error Reading IPD Response",e);
		}
        
        return spResponseParser;
    }

	/**
	 * Post data to the Service Provider 
	 * 
	 * @param spResponse - The response that contains data for the Service Provider
	 * @throws ConnectionException
	 */
    private void ServiceProviderRequest(IResponseParser spResponse) throws ConnectionException
    {
    	PostData formData = spResponse.getPostData();

        HttpPost request;
		try {
			request = HttpPost.create(StringEscapeUtils.unescapeHtml(spResponse.getPostUrl()), "application/x-www-form-urlencoded");
			request.setData(formData.toString());
		} catch (MalformedURLException e) {
			throw new ConnectionException("Invalid URL for Service Provider Request", e);
		}
		
        HttpResponse respponse = new HttpClient().execute(request);
        respponse.close();        
    }
    
    private final Map<String, HttpClient> _requests = new HashMap<String, HttpClient>();
	@Override
	public OutputStream beginRequest(String path, String contentType) throws ConnectionException {

		HttpPost request;
		try {
			request = HttpPost.create(_v1Url + path, contentType);
		} catch (MalformedURLException e) {
			throw new ConnectionException("Invalid URL " + _v1Url + path, e);
		}
		HttpClient client = new HttpClient();
		_requests.put(path, client);
		
		try {
			return client.beginRequest(request).getOutputStream();
		} catch (IOException e) {
			throw new ConnectionException("Cannot begin the request for " + _v1Url + path, e);
		}		
	}

	@Override
	public InputStream endRequest(String path) throws ConnectionException {
		
		HttpClient client = _requests.get(path); 
		try {
			return client.endRequest().getInputStream();
		} catch (IOException e) {
			throw new ConnectionException("Cannot get response stream for " + path, e);
		}		
	}

	@Override
	public Reader getData() throws ConnectionException {
		return getData("");
	}

	@Override
	public Reader getData(String path) throws ConnectionException {
		
    	HttpGet request;
		try {
			request = HttpGet.create(_v1Url + path);
		} catch (MalformedURLException e) {
			throw new ConnectionException("Invalid URL " + _v1Url + path, e);
		}
		
    	HttpResponse response = new HttpClient().execute(request);
    	if(isSamlResponse(response.getContentType()))
    	{
    		response.close();
    		
			authenticate();
    		return getData(path);
    	}

    	try {
			return new InputStreamReader(response.getInputStream());
		} catch (IOException e) {
			throw new ConnectionException("Error processing result from URL "
					+ path, e);
		}
	}

	private boolean isSamlResponse(String contentType) {
		return contentType.startsWith("text/html");
	}

	@Override
	public Reader sendData(String path, String data) throws ConnectionException {
		HttpPost request;
		try {
			request = HttpPostXmlContent.create(_v1Url + path, data);
		} catch (MalformedURLException e) {
			throw new ConnectionException("Invalid URL " + _v1Url + path, e);
		}

    	HttpResponse response = new HttpClient().execute(request);
    	try {
			return new InputStreamReader(response.getInputStream());
		} catch (IOException e) {
			throw new ConnectionException("Error processing result from URL "
					+ path, e);
		}
	}

	/**
	 * Authenticator
	 * @author Jerry D. Odenwelder Jr.
	 *
	 */
	private class Credentials extends Authenticator {

		PasswordAuthentication _value;

		@Override
		protected PasswordAuthentication getPasswordAuthentication() {
			return _value;
		}

		Credentials(String userName, String password) {
			if (null == password) {
				_value = new PasswordAuthentication(userName, "".toCharArray());
			} else {
				_value = new PasswordAuthentication(userName, password
						.toCharArray());
			}
		}

		@Override
		public String toString() {
			return _value.getUserName() + ":"
					+ String.valueOf(_value.getPassword());
		}
	}
}
